server/etcdserver/api/etcdhttp/health.go

```diff
@@ -118,78 +133,251 @@ type Health struct {
 	Reason string `json:"reason"`
 }
 
-type AlarmSet map[string]struct{}
-
-func getExcludedAlarms(r *http.Request) (alarms AlarmSet) {
-	alarms = make(map[string]struct{}, 2)
-	alms, found := r.URL.Query()["exclude"]
+func getQuerySet(r *http.Request, query string) StringSet {
+	querySet := make(map[string]struct{})
+	qs, found := r.URL.Query()[query]
 	if found {
-		for _, alm := range alms {
-			if len(alm) == 0 {
+		for _, q := range qs {
+			if len(q) == 0 {
 				continue
 			}
-			alarms[alm] = struct{}{}
+			querySet[q] = struct{}{}
 		}
 	}
-	return alarms
+	return querySet
 }
 
 func getSerializableFlag(r *http.Request) bool {
 	return r.URL.Query().Get("serializable") == "true"
 }
 
-// TODO: etcdserver.ErrNoLeader in health API
+type HealthCheck func(ctx context.Context) error
+
+type CheckRegistry struct {
+	path           string
+	failCounter    prometheus.Counter
+	successCounter prometheus.Counter
+
+	checks map[string]HealthCheck
+	// the list of checkNames are needed to ensure the order of checks.
+	checkNames []string
+}
+
+func DefaultCheckRegistry(path string) *CheckRegistry {
+	reg := CheckRegistry{path: path, checks: make(map[string]HealthCheck), failCounter: healthFailed, successCounter: healthSuccess}
+	return &reg
+}
+
+// installHealthEndpoints installs endpoints for /health
+func installHealthEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
```

Comment:
I see you did a full refactor of the existing health endpoint to the CheckRegistry. This is good change, however it makes the PR harder to review as reviewer not only needs to check the new health, but also validate if there were not unintentional changes to the old health. 

Could you consider moving the refactor to a separate PR?